File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py

from collections import deque
def solution(n, wires):
    answer = float('inf')
    # answer =-1
    
    for i in range(len(wires)):
        temp_wires = wires[:i] + wires[i+1:]

        graph = [[] for _ in range(n+1)]
        for a,b in temp_wires:
            graph[a].append(b)
            graph[b].append(a)
            
        visited_list = [False] * (n+1)
        visited_list[1] = True
        q=deque([1])
        cnt = 1
            
        while q:
            e_top=q.popleft()
            for link_top in graph[e_top]:
                if not visited_list[link_top]:
                    visited_list[link_top] = True
                    q.append(link_top)
                    cnt +=1
        diff = abs(n-cnt-cnt)
        answer = min(answer,diff)
            
    return answer

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] wires) {
        int answer = Integer.MAX_VALUE;
        
        
        List<List<Integer>> graph = new ArrayList<>();
        for(int i=0; i<=n; i++){
            graph.add(new ArrayList<>());
        }
        
        for(int[] wire : wires){
            int a = wire[0];
            int b = wire[1];
            
            graph.get(a).add(b);
            graph.get(b).add(a);
        }
        
        
        for(int[] wire : wires){
            int start_wire = wire[0];
            int end_wire = wire[1];
            
            int count = bfsCount(graph, start_wire, end_wire,n);
            int diff = Math.abs((n-count) - count);
            answer = Math.min(answer, diff);
        }
        
        
        
        
        return answer;
    }
    
    private int bfsCount(List<List<Integer>> graph,int a,int b, int n){
        boolean[] visited = new boolean[n+1];
        Queue<Integer> queue = new LinkedList<>();
        
        queue.offer(a);
        visited[a] = true;
        int count = 1;
        
        while(!queue.isEmpty()){
            int x = queue.poll();
            
            for(int num : graph.get(x)){
                if((x==a && num ==b) || (x==b && num==a)) continue;
                
                if(!visited[num]){
                    queue.offer(num);
                    visited[num]=true;
                    count++;
                }
            }
            
        }
        
        return count;
    }
}